lib/bolt/module_installer.rb

```diff
@@ -13,45 +17,49 @@ def initialize(outputter, pal)
 
     # Adds a single module to the project.
     #
-    def add(name, modules, puppetfile_path, moduledir, config_path)
-      require 'bolt/puppetfile'
-
-      @outputter.print_message("Adding module #{name} to project\n\n")
-
-      # If the project configuration file already includes this module,
-      # exit early.
-      puppetfile  = Bolt::Puppetfile.new(modules)
-      new_module  = Bolt::Puppetfile::Module.from_hash('name' => name)
-
-      if puppetfile.modules.include?(new_module)
-        @outputter.print_action_step(
-          "Project configuration file #{config_path} already includes module #{new_module}. Nothing to do."
+    def add(name, specs, puppetfile_path, moduledir, config_path)
+      specs = Specs.new(specs)
+
+      # Exit early if project config already includes a spec with this name.
+      if specs.include?(name)
+        @outputter.print_message(
+          "Project configuration file #{config_path} already includes specification with name "\
+          "#{name}. Nothing to do."
         )
         return true
       end
 
+      @outputter.print_message("Adding module #{name} to project\n\n")
+
+      # Create a Puppetfile object. If there isn't an existing Puppetfile,
+      # this won't include any resolved modules.
+      existing_puppetfile = Puppetfile.parse(puppetfile_path)
```

Comment:
`new_specs` and `existing_specs` or `old_specs` might also be better names? But up to you

Reply:
This change would make `existing_puppetfile` inaccessible, which is needed as a parameter to `print_puppetfile_diff` after resolving.

Reply:
Couldn't `print_puppetfile_diff` just handle the old puppetfile not existing - probably it could just default to making `old` an empty hash if it gets nil from the old puppetfile.